pkg/distributor/query.go

```diff
@@ -0,0 +1,187 @@
+func (d *Distributor) queryIngesters(ctx context.Context, replicationSet ring.ReplicationSet, req *client.QueryRequest) (model.Matrix, error) {
+	// Fetch samples from multiple ingesters in parallel, using the replicationSet
+	// to deal with consistency.
+	results, err := replicationSet.Do(ctx, func(ing *ring.IngesterDesc) (interface{}, error) {
+		client, err := d.ingesterPool.GetClientFor(ing.Addr)
+		if err != nil {
+			return nil, err
+		}
+
+		resp, err := client.(ingester_client.IngesterClient).Query(ctx, req)
+		ingesterQueries.WithLabelValues(ing.Addr).Inc()
+		if err != nil {
+			ingesterQueryFailures.WithLabelValues(ing.Addr).Inc()
+			return nil, err
+		}
+
+		return ingester_client.FromQueryResponse(resp), nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Merge the results into a single matrix.
+	fpToSampleStream := map[model.Fingerprint]*model.SampleStream{}
+	for _, result := range results {
+		for _, ss := range result.(model.Matrix) {
+			fp := ss.Metric.Fingerprint()
+			mss, ok := fpToSampleStream[fp]
+			if !ok {
+				mss = &model.SampleStream{
+					Metric: ss.Metric,
+				}
+				fpToSampleStream[fp] = mss
+			}
+			mss.Values = util.MergeSampleSets(mss.Values, ss.Values)
+		}
+	}
+	result := model.Matrix{}
+	for _, ss := range fpToSampleStream {
+		result = append(result, ss)
+	}
+
+	return result, nil
+}
+
+// queryIngesterStream queries the ingesters using the new streaming API.
+func (d *Distributor) queryIngesterStream(ctx context.Context, replicationSet ring.ReplicationSet, req *client.QueryRequest) ([]client.TimeSeriesChunk, error) {
+	// Fetch samples from multiple ingesters
+	results, err := replicationSet.Do(ctx, func(ing *ring.IngesterDesc) (interface{}, error) {
+		client, err := d.ingesterPool.GetClientFor(ing.Addr)
+		if err != nil {
+			return nil, err
+		}
+
+		stream, err := client.(ingester_client.IngesterClient).QueryStream(ctx, req)
+		ingesterQueries.WithLabelValues(ing.Addr).Inc()
+		if err != nil {
+			ingesterQueryFailures.WithLabelValues(ing.Addr).Inc()
+			return nil, err
+		}
+		defer stream.CloseSend()
+
+		var result []*ingester_client.QueryStreamResponse
+		for {
+			series, err := stream.Recv()
+			if err == io.EOF {
+				break
+			} else if err != nil {
+				return nil, err
+			}
+			result = append(result, series)
+		}
+		return result, nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	hashToSeries := map[model.Fingerprint]ingester_client.TimeSeriesChunk{}
+	for _, result := range results {
+		for _, response := range result.([]*ingester_client.QueryStreamResponse) {
+			for _, series := range response.Timeseries {
+				hash := ingester_client.FromLabelPairs(series.Labels).FastFingerprint()
+				existing := hashToSeries[hash]
+				existing.Labels = series.Labels
+				existing.Chunks = append(existing.Chunks, series.Chunks...)
+				hashToSeries[hash] = existing
+			}
+		}
+	}
+	result := make([]client.TimeSeriesChunk, 0, len(hashToSeries))
+	for hash := range hashToSeries {
```

Comment:
```
for _, series := range hashToSeries {
  result = append(result, series)
}
```

Is this not done this way to avoid copying? Could you add a comment saying so?

Reply:
It appears not:

`./query.go:158:32: invalid type assertion: results.([][]*client.QueryStreamResponse) (non-interface type []interface {} on left)`